File: client/shared/src/commonMain/kotlin/com/chefstol/app/core/mvi/ScreenState.kt
package com.chefstol.app.core.mvi

// Direct reflection of 09_Логики/LOGIC-005_Паттерн-состояний-экрана.md — do not add states
// beyond what that document defines (no separate "past"/"upcoming" screen state, etc.).
sealed interface ScreenState<out T> {
    data object Loading : ScreenState<Nothing>
    data class Content<T>(val value: T, val refreshing: Boolean = false) : ScreenState<T>
    data class Empty(val reason: String) : ScreenState<Nothing>
    data class Error(val message: String) : ScreenState<Nothing>
}

// Actions (createBooking/cancelBooking/submitRating in later iterations) never have their own
// Loading/Content/Empty/Error — only a busy flag on the triggering control (LOGIC-005 "Обратная
// связь по действию").
enum class ActionStatus { Idle, Submitting }

File: client/src/core/screenState.ts
// Direct reflection of 5-mobile-app-spec/09_Логики/LOGIC-005_Паттерн-состояний-экрана.md —
// not a generic "just in case" abstraction. Loading is a full-screen skeleton shown only when
// there's nothing to display yet; Refreshing keeps prior content visible during pull-to-refresh
// and never falls back to Error (LOGIC-005 AC-007).
export type ScreenState<T> =
  | { kind: 'loading' }
  | { kind: 'content'; value: T; refreshing: boolean }
  | { kind: 'empty'; reason: string }
  | { kind: 'error'; message: string }

export const ScreenState = {
  loading: <T>(): ScreenState<T> => ({ kind: 'loading' }),
  content: <T>(value: T, refreshing = false): ScreenState<T> => ({ kind: 'content', value, refreshing }),
  empty: <T>(reason: string): ScreenState<T> => ({ kind: 'empty', reason }),
  error: <T>(message: string): ScreenState<T> => ({ kind: 'error', message }),
}
